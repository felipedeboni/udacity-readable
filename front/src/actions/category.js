export const FETCH_CATEGORIES = 'FETCH_CATEGORIES';
export const FETCH_CATEGORIES_SUCCESS = 'FETCH_CATEGORIES_SUCCESS';
export const FETCH_CATEGORIES_FAIL = 'FETCH_CATEGORIES_FAIL';

export const fetchCategories = () => {
  return {
    type: FETCH_CATEGORIES,
    payload: {
      request: {
        url: '/categories'
      }
    }
  };
};
