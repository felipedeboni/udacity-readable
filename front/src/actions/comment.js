import uuid from 'utils/uuid';

// GET /posts/:id/comments
// Get all the comments for a single post.
export const FETCH_COMMENTS_BY_POST = 'FETCH_COMMENTS_BY_POST';
export const FETCH_COMMENTS_BY_POST_SUCCESS = 'FETCH_COMMENTS_BY_POST_SUCCESS';
export const FETCH_COMMENTS_BY_POST_FAIL = 'FETCH_COMMENTS_BY_POST_FAIL';
export const fetchCommentsByPost = postId => {
  return {
    type: FETCH_COMMENTS_BY_POST,
    payload: {
      request: {
        url: `/posts/${postId}/comments`
      }
    }
  };
};

// GET /comments/:id
// Get the details for a single comment.
export const FETCH_COMMENT = 'FETCH_COMMENT';
export const FETCH_COMMENT_SUCCESS = 'FETCH_COMMENT_SUCCESS';
export const FETCH_COMMENT_FAIL = 'FETCH_COMMENT_FAIL';
export const fetchComment = id => {
  return {
    type: FETCH_COMMENT,
    payload: {
      request: {
        url: `/comments/${id}`
      }
    }
  };
};

// POST /comments
// Add a comment to a post.
export const ADD_COMMENT = 'ADD_COMMENT';
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS';
export const ADD_COMMENT_FAIL = 'ADD_COMMENT_FAIL';
export const addComment = (parentId, values) => {
  const { author, body } = values;
  const data = {
    id: uuid(),
    timestamp: Date.now(),
    parentId,
    author,
    body
  };

  return {
    type: ADD_COMMENT,
    payload: {
      request: {
        method: 'POST',
        url: '/comments',
        data
      }
    }
  };
};

// PUT /comments/:id
// Edit the details of an existing comment.
export const EDIT_COMMENT = 'EDIT_COMMENT';
export const EDIT_COMMENT_SUCCESS = 'EDIT_COMMENT_SUCCESS';
export const EDIT_COMMENT_FAIL = 'EDIT_COMMENT_FAIL';
export const editComment = (id, data) => {
  return {
    type: EDIT_COMMENT,
    payload: {
      request: {
        method: 'PUT',
        url: `/comments/${id}`,
        data
      }
    }
  };
};

// DELETE /comments/:id
// Sets a comment's deleted flag to true.
export const REMOVE_COMMENT = 'REMOVE_COMMENT';
export const REMOVE_COMMENT_SUCCESS = 'REMOVE_COMMENT_SUCCESS';
export const REMOVE_COMMENT_FAIL = 'REMOVE_COMMENT_FAIL';
export const removeComment = id => {
  return {
    type: REMOVE_COMMENT,
    payload: {
      request: {
        method: 'DELETE',
        url: `/comments/${id}`
      }
    }
  };
};

// POST /comments/:id
// Used for voting on a comment.
export const VOTE_COMMENT = 'VOTE_COMMENT';
export const VOTE_COMMENT_SUCCESS = 'VOTE_COMMENT_SUCCESS';
export const VOTE_COMMENT_FAIL = 'VOTE_COMMENT_FAIL';
const vote = (id, type) => {
  return {
    type: VOTE_COMMENT,
    payload: {
      request: {
        method: 'POST',
        url: `/comments/${id}`,
        data: {
          option: type
        }
      }
    }
  };
};

export const voteCommentUp = id => {
  return vote(id, 'upVote');
};

export const voteCommentDown = id => {
  return vote(id, 'downVote');
};
