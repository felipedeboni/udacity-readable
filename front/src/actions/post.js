import uuid from 'utils/uuid';

// GET /posts
// Get all of the posts. Useful for the main page when no category is selected.
export const FETCH_ALL_POSTS = 'FETCH_ALL_POSTS';
export const FETCH_ALL_POSTS_SUCCESS = 'FETCH_ALL_POSTS_SUCCESS';
export const FETCH_ALL_POSTS_FAIL = 'FETCH_ALL_POSTS_FAIL';
export const fetchAllPosts = () => {
  return {
    type: FETCH_ALL_POSTS,
    payload: {
      request: {
        url: `/posts`
      }
    }
  };
};

// GET /:category/posts
// Get all of the posts for a particular category.
export const FETCH_POSTS_BY_CATEGORY = 'FETCH_POSTS_BY_CATEGORY';
export const FETCH_POSTS_BY_CATEGORY_SUCCESS =
  'FETCH_POSTS_BY_CATEGORY_SUCCESS';
export const FETCH_POSTS_BY_CATEGORY_FAIL = 'FETCH_POSTS_BY_CATEGORY_FAIL';
export const fetchPostsByCategory = categoryId => {
  return {
    type: FETCH_POSTS_BY_CATEGORY,
    payload: {
      request: {
        url: `/${categoryId}/posts`
      }
    }
  };
};

// GET /posts/:id
// Get the details of a single post.
export const FETCH_POST = 'FETCH_POST';
export const FETCH_POST_SUCCESS = 'FETCH_POST_SUCCESS';
export const FETCH_POST_FAIL = 'FETCH_POST_FAIL';
export const fetchPost = id => {
  return {
    type: FETCH_POST,
    payload: {
      request: {
        url: `/posts/${id}`
      }
    }
  };
};

// POST /posts
// Add a new post.
export const ADD_POST = 'ADD_POST';
export const ADD_POST_SUCCESS = 'ADD_POST_SUCCESS';
export const ADD_POST_FAIL = 'ADD_POST_FAIL';
export const addPost = ({ category, author, title, body }) => {
  const data = {
    id: uuid(),
    timestamp: Date.now(),
    category,
    author,
    title,
    body
  };

  return {
    type: ADD_POST,
    payload: {
      request: {
        method: 'POST',
        url: '/posts',
        data
      }
    }
  };
};

// PUT /posts/:id
// Edit the details of an existing post.
export const EDIT_POST = 'EDIT_POST';
export const EDIT_POST_SUCCESS = 'EDIT_POST_SUCCESS';
export const EDIT_POST_FAIL = 'EDIT_POST_FAIL';
export const editPost = (id, data) => {
  return {
    type: EDIT_POST,
    payload: {
      request: {
        method: 'PUT',
        url: `/posts/${id}`,
        data
      }
    }
  };
};

// DELETE /posts/:id
// Sets the deleted flag for a post to 'true'.
// Sets the parentDeleted flag for all child comments to 'true'.
export const REMOVE_POST = 'REMOVE_POST';
export const REMOVE_POST_SUCCESS = 'REMOVE_POST_SUCCESS';
export const REMOVE_POST_FAIL = 'REMOVE_POST_FAIL';
export const removePost = id => {
  return {
    type: REMOVE_POST,
    payload: {
      id: id,
      request: {
        method: 'DELETE',
        url: `/posts/${id}`
      }
    }
  };
};

// POST /posts/:id
// Used for voting on a post.
export const VOTE_POST = 'VOTE_POST';
export const VOTE_POST_SUCCESS = 'VOTE_POST_SUCCESS';
export const VOTE_POST_FAIL = 'VOTE_POST_FAIL';
const vote = (id, type) => {
  return {
    type: VOTE_POST,
    payload: {
      request: {
        method: 'POST',
        url: `/posts/${id}`,
        data: {
          option: type
        }
      }
    }
  };
};

export const votePostUp = id => {
  return vote(id, 'upVote');
};

export const votePostDown = id => {
  return vote(id, 'downVote');
};
