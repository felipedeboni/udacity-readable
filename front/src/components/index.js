export { default as App } from './App/App';

export { default as Header } from './Header/Header';

export { default as CommentForm } from './CommentForm/CommentForm';
export { default as CommentsList } from './CommentsList/CommentsList';
export { default as CommentListItem } from './CommentListItem/CommentListItem';

export { default as Post } from './Post/Post';
export { default as PostForm } from './PostForm/PostForm';
export { default as PostsList } from './PostsList/PostsList';
export { default as PostListItem } from './PostListItem/PostListItem';

export { default as Score } from './Score/Score';

export { default as PageNotFound } from './PageNotFound/PageNotFound';
