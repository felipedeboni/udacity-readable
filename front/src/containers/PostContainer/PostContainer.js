import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchPost, removePost, FETCH_POST_FAIL } from 'actions/post';

class PostContainer extends Component {
  componentDidMount() {
    const { postId, post, fetch } = this.props;

    if (!post) {
      fetch(postId)
        .then(this.afterFetch)
        .catch(this.afterFetch);
    }
  }

  afterFetch = response => {
    const { history, match: { params: { category } } } = this.props;
    const is404 =
      response.type === FETCH_POST_FAIL ||
      !response.payload ||
      !response.payload.data.id ||
      response.payload.data.category !== category;
    if (is404) {
      history.replace('/404');
    }
  };

  render() {
    const { component, post, remove, history } = this.props;

    if (post) {
      return React.createElement(
        component,
        {
          post,
          history,
          removeHandler: remove
        },
        null
      );
    }

    return null;
  }
}

const mapProps = (state, ownProps) => {
  return {
    post: ownProps.post || state.posts[ownProps.postId]
  };
};

const mapDispatch = dispatch => {
  return {
    fetch: postId => dispatch(fetchPost(postId)),
    remove: postId => dispatch(removePost(postId))
  };
};

export default connect(mapProps, mapDispatch)(PostContainer);
