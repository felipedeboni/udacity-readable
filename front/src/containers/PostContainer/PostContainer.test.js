import React from 'react';
import * as PostActions from 'actions/post';
import PostContainer from './PostContainer';
import { Post } from 'components';

describe('PostContainer', () => {
  const render = () => {
    const [post] = global.mock.posts;
    const state = {
      posts: global.mock.postsObject
    };

    const { component, store } = global.shallowWithStore(
      <PostContainer
        postId={post.id}
        match={{ params: { category: post.category } }}
        component={Post}
      />,
      state
    );

    return {
      component,
      store,
      post
    };
  };

  it('renders without crashing', () => {
    const { component } = render();

    expect(component.dive()).toMatchSnapshot();
  });

  it('fetches the post', () => {
    const [post] = global.mock.posts;
    const { component, store } = global.shallowWithStore(
      <PostContainer
        id={post.id}
        match={{ params: { category: post.category } }}
        component={Post}
        history={{ push: jest.fn(), replace: jest.fn() }}
      />,
      { posts: {} }
    );

    expect(component.dive()).toMatchSnapshot();
    expect(
      store.getActions().find(_ => _.type === PostActions.FETCH_POST)
    ).toBeDefined();
  });

  it('dispatch removePost', () => {
    const { component, store } = render();
    const form = component;

    form.props().remove({});
    expect(
      store.getActions().find(_ => _.type === PostActions.REMOVE_POST)
    ).toBeDefined();
  });
});
