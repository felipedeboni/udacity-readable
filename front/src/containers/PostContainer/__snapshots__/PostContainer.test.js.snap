// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostContainer fetches the post 1`] = `""`;

exports[`PostContainer renders without crashing 1`] = `
<Post
  post={
    Object {
      "author": "thingtwo",
      "body": "Everyone says so after all.",
      "category": "react",
      "commentCount": 2,
      "deleted": false,
      "id": "8xf0y6ziyjabvozdd253nd",
      "timestamp": 1467166872634,
      "title": "Udacity is the best place to learn React",
      "voteScore": 6,
    }
  }
  removeHandler={[Function]}
/>
`;
